### PR TITLE
Build/upgrade tools

### PR DESCRIPTION
**Pull request by soofstad**
A bit big. But hard to avoid since upgrading one thing causes other packages to be incompatible, leading to those needing an upgrade as well.

Suggest looking through each commit to see which kind of changes was made.

## What does this pull request change?
* bump openapi generator
* bump FastAPI and gunicorn
* fix ruff linting violations
* fix mypy errors
* replace black, pycln, isort, flake8, bandit, with RUFF
*  bump python -> 3.12

## Why is this pull request needed?
- Toolchain was getting dated

### CODE REVIEW

File: src/common/tree/tree_node.py [hunk 1]
```diff
@@ -138,7 +135,7 @@ def __repr__(self):
         return f"Key: '{self.key}', Type: '{self.type}', Node_ID: '{self.node_id}'"
 
     def show_tree(self, level=0):
-        print("%s%s" % ("." * level, self))
+        print("{}{}".format("." * level, self))
```
Comment (pbullhove) [inline src/common/tree/tree_node.py:138]:
```suggestion
print(f"{'.' * level}{self}")
```

Comment (pbullhove) [reply src/common/tree/tree_node.py:138]: f strings are cleaner

File: src/common/tree/tree_node_serializer.py [hunk 2]
```diff
@@ -183,7 +173,7 @@ def tree_node_from_dict(
             if not isinstance(children, list):
                 raise ValueError(
                     f"The attribute '{child_attribute.name}' on blueprint '{node.type}' "
-                    + f"should be a list, but was '{str(type(children))}'"
+                    + f"should be a list, but was '{type(children)!s}'"
```
Comment (pbullhove) [inline src/common/tree/tree_node_serializer.py:176]: Pretty sure the !s is unecessary here.
Tested this in playground, the !s makes no difference in output.

File: src/common/tree/tree_node_serializer.py [hunk 3]
```diff
@@ -225,7 +215,7 @@ def tree_node_from_dict(
             if not isinstance(attribute_data, dict):
                 raise ValidationException(
                     f"The attribute '{child_attribute.name}' on blueprint '{node.type}' "
-                    + f"should be a dict, but was '{str(type(attribute_data))}'"
+                    + f"should be a dict, but was '{type(attribute_data)!s}'"
```
Comment (pbullhove) [inline src/common/tree/tree_node_serializer.py:218]: Duplicate:

Pretty sure the !s is unecessary here.
Tested this in playground, the !s makes no difference in output.

File: src/domain_classes/dimension.py [hunk 4]
```diff
@@ -10,8 +10,8 @@ def _get_data_type_from_dmt_type(
         return type_enum.to_py_type()
     except ValueError:
         return attribute_type
-    except Exception as error:
-        raise Exception(f"Something went wrong trying to fetch data type: {error}")
+    except Exception as ex:
```
Comment (pbullhove) [inline src/domain_classes/dimension.py:13]: should we not avoid catching Exception, as it is way too general, and will catch even "ctrl+c" force quit etc. 

Also I think the exception message "data type: {ex}" does not quite make sense.

File: src/storage/internal/data_source_repository.py [hunk 5]
```diff
@@ -56,12 +52,10 @@ def validate_repository(repo: dict):
                     raise BadRequestException(
                         f"The database named '{repo['database']}' " "is a system reserved database name."
                     )
-        except KeyError as e:
-            raise KeyError(e)
         except Exception as error:
```
Comment (pbullhove) [inline src/storage/internal/data_source_repository.py:55]:
```suggestion
        except Exception as ex:
```

File: src/storage/internal/data_source_repository.py [hunk 6]
```diff
@@ -56,12 +52,10 @@ def validate_repository(repo: dict):
...
-            raise BadRequestException(error)
+            raise BadRequestException(error) from error
```
Comment (pbullhove) [inline src/storage/internal/data_source_repository.py:56]:
```suggestion
            raise BadRequestException(ex) from ex
```

File: src/storage/repositories/mongo.py [hunk 7]
```diff
@@ -40,23 +39,23 @@ def get(self, uid: str) -> dict:
...
         return result
 
-    def add(self, uid: str, document: Dict) -> bool:
+    def add(self, uid: str, document: dict) -> bool:
         document["_id"] = uid
         try:
             return self.handler[self.collection].insert_one(document).acknowledged
-        except DuplicateKeyError:
-            raise BadRequestException
+        except DuplicateKeyError as error:
```
Comment (pbullhove) [inline src/storage/repositories/mongo.py:46]:
```suggestion
        except DuplicateKeyError as ex:
```

File: src/storage/repositories/mongo.py [hunk 8]
```diff
@@ -40,23 +39,23 @@ def get(self, uid: str) -> dict:
...
+            raise BadRequestException from error
```
Comment (pbullhove) [inline src/storage/repositories/mongo.py:47]:
```suggestion
            raise BadRequestException from ex
```

File: src/storage/repositories/mongo.py [hunk 9]
```diff
@@ -71,10 +70,10 @@ def update_blob(self, uid: str, blob: bytearray):
                 sleep(2)
                 if attempts > 2:
                     raise error
-            except gridfs.errors.FileExists:
+            except gridfs.errors.FileExists as error:
```
Comment (pbullhove) [inline src/storage/repositories/mongo.py:73]:
```suggestion
            except gridfs.errors.FileExists as ex:
```

File: src/storage/repositories/mongo.py [hunk 10]
```diff
@@ -71,10 +70,10 @@ def update_blob(self, uid: str, blob: bytearray):
...
                 message = f"Blob file with id '{uid}' already exists"
                 logger.warning(message)
-                raise BadRequestException(message=message)
+                raise BadRequestException(message=message) from error
```
Comment (pbullhove) [inline src/storage/repositories/mongo.py:76]:
```suggestion
                raise BadRequestException(message=message) from ex
```

File: src/tests/bdd/steps/api.py [hunk 11]
```diff
@@ -122,19 +122,19 @@ def step_user_logout(context):
...
-def step_impl(context):
+def step_PAT_is_expired(context):
     sleep(2)  # Give the PAT some time to expire
     context.response = context.test_client.get("/api/whoami", headers=context.headers)
     assert context.response.status_code == 401
 
 
 @step("the PAT is revoked")
-def step_impl(context):
+def step_PAT_is_revoked(context):
```
Comment (pbullhove) [inline src/tests/bdd/steps/api.py:137]: nice!

**Review by pbullhove — APPROVED** [review]
Very nice, good work

File: src/common/tree/tree_node_serializer.py (hunk 2 above)
Comment (soofstad) [reply src/common/tree/tree_node_serializer.py:176]: removed

File: src/common/tree/tree_node.py (hunk 1 above)
Comment (soofstad) [reply src/common/tree/tree_node.py:138]: Agreed

File: src/common/entity/validators.py [hunk 12]
```diff
@@ -111,7 +110,7 @@ def _validate_entity(
 
     if implementation_mode == "exact":
         if keys_not_in_blueprint := [
-            key for key in entity.keys() if key not in (blueprint.get_attribute_names() + ["_id"])
+            key for key in entity.keys() if key not in ([*blueprint.get_attribute_names(), "_id"])
```
[the remark this answers is not in the document]
Comment (soofstad) [reply src/common/entity/validators.py:113]: just that the formater did not remove it after refactoring. Good catch :+1:

File: pyproject.toml [hunk 13]
```diff
@@ -41,19 +38,32 @@ pythonpath = [
...
+    "S",  # Bandit
+    "C",  # flake8-comprehensions
+    "B",  # flake8-bugbear
+    "UP", # automatically upgrade syntax for newer versions of the language
+    "RUF", # ruff specific rules
+]
+
+ignore = [
+    "W191", # conflict with ruff formatter
+    "E501",  # line too long, handled by black,
```
[the remark this answers is not in the document]
Comment (soofstad) [reply pyproject.toml:61]: Reworded comment

File: src/domain_classes/dimension.py (hunk 4 above)
Comment (soofstad) [reply src/domain_classes/dimension.py:13]: Agreed, but don't want to touch logic with this PR :)

**Pull request merged**